packages/core/src/lib/connection_manager.ts

```diff
@@ -0,0 +1,381 @@
+    relay?: IRelay
+  ): void {
+    const onPeerConnect = (
+      evt: CustomEvent<Connection>,
+      keepAliveOptions: KeepAliveOptions,
+      relay?: IRelay
+    ): void => {
+      {
+        this.keepAliveManager.start(
+          evt.detail.remotePeer,
+          this.libp2pComponents.ping.bind(this),
+          keepAliveOptions,
+          relay
+        );
+      }
+    };
+
+    this.libp2pComponents.connectionManager.addEventListener(
+      "peer:connect",
+      (evt) => onPeerConnect(evt, keepAliveOptions, relay)
+    );
+  }
+
+  private startPeerDisconnectionListener(): void {
+    const onPeerDisconnect = () => {
+      return (evt: CustomEvent<Connection>): void => {
+        this.keepAliveManager.stop(evt.detail.remotePeer);
+      };
+    };
+
+    // TODO: ensure that these following issues are updated and confirmed
+    /**
+     * NOTE: Event is not being emitted on closing nor losing a connection.
+     * @see https://github.com/libp2p/js-libp2p/issues/939
+     * @see https://github.com/status-im/js-waku/issues/252
+     *
+     * >This event will be triggered anytime we are disconnected from another peer,
+     * >regardless of the circumstances of that disconnection.
+     * >If we happen to have multiple connections to a peer,
+     * >this event will **only** be triggered when the last connection is closed.
+     * @see https://github.com/libp2p/js-libp2p/blob/bad9e8c0ff58d60a78314077720c82ae331cc55b/doc/API.md?plain=1#L2100
+     */
+    this.libp2pComponents.connectionManager.addEventListener(
+      "peer:disconnect",
+      onPeerDisconnect
+    );
+  }
+
+  /**
+   * Checks if the peer is dialable based on the following conditions:
+   * 1. If the peer is a bootstrap peer, it is only dialable if the number of current bootstrap connections is less than the max allowed.
+   * 2. If the peer is not a bootstrap peer
+   */
+  private async shouldDialPeer(peerId: PeerId): Promise<boolean> {
+    const { maxBootstrapPeersAllowed = DEFAULT_MAX_BOOTSTRAP_PEERS_ALLOWED } =
+      this.options;
+
+    const isConnected =
+      this.libp2pComponents.connectionManager.getConnections(peerId).length > 0;
+
+    if (isConnected) return false;
+
+    const isBootstrap = (await this.getTagNamesForPeer(peerId)).some(
+      (tagName) => tagName === Tags.BOOTSTRAP
+    );
+
+    if (isBootstrap) {
+      const currentBootstrapConnections =
+        this.libp2pComponents.connectionManager
+          .getConnections()
+          .filter((conn) => {
+            conn.tags.find((name) => name === Tags.BOOTSTRAP);
+          }).length;
+      if (currentBootstrapConnections < maxBootstrapPeersAllowed) return true;
+    } else {
+      return true;
+    }
+
+    return false;
+  }
+
+  /**
+   * Fetches the updated state of the libp2p peerStore and connectionManager
+   * Also logs the updated state
+   * @returns UpdatedStates
+   */
+  private async fetchUpdatedState(): Promise<UpdatedStates> {
```

Comment:
I think we should reconsider this approach in general. Please, correct me if I am wrong in my reasoning. 

Right now we have this callback being called on interval basis and it does couple of things such as:
- iterating through all peers and gathering info about them;
- determines bootstrapped and not bootstrapped connections;

In my opinion we should split this function into two:
- one that runs through `peers` on initialisation of class (could be runService)
- one that subscribes to `peer` event from `PeerStorage` https://github.com/libp2p/js-libp2p-interfaces/blob/29d6923bbd9bfb0335db8b8c3949d69bb9d4c2db/packages/interface-peer-store/src/index.ts#L332

It seems it should be possible.

Reply:
That's what I want to prevent. If we can go with subscription based approach instead of polling it is almost always better. 

So, based on your knowledge, is it possible to refactor this approach into two stages when one is for initialisation (getting all peers and iterating over them) and the other one for keeping state up to date by subscribing to the event?

To me it seems we can do that and subscribe to `peer` event: https://github.com/libp2p/js-libp2p-interfaces/blob/29d6923bbd9bfb0335db8b8c3949d69bb9d4c2db/packages/interface-peer-store/src/index.ts#L332

Reply:
So in case we already have a listener to this event we can just add this logic in there or add additional listener. To me it seems this is better that having interval.

Reply:
> having one source of truth is always preferred vs maintaining states manually (redundantly)

Then use the peer store as a source of truth and use tagging to mark the peers instead of having the `UpdatedStates` object.